Add support for programming from intel hex files

g-oikonomou opened the pull request:
This will need rebased, but feel free to have a look at the feature. It is implemented in the last 4 commits of this pull.

Requires `IntelHex` and `python-magic` in site-packages.

With those packages available, we auto-detect the firmware type.

If those packages are not available, support for .hex will be disabled gracefully and the script will still work the way it does ~~not~~ for bin files.

JelmerT commented:
This looks like a very nice feature, and will make the script more useful!

Let me play around with it a little bit (and pull the 26xx support first), and we'll try to get it in there.

g-oikonomou commented:
This is rebased and ready for consideration

JelmerT commented on README.md:
```diff
@@ -12,6 +12,14 @@ To communicate with the uart port of the SoC you need a usb to serial converter:
+
+If you want to be able to program your device from an intel hex file, you will also need to install the following:
+
+* IntelHex: https://pypi.python.org/pypi/IntelHex
```
Maybe add 'pip install intelhex' here? One less link to visit for the programmer that's less familiar with python.

JelmerT commented on README.md:
```diff
@@ -12,6 +12,14 @@ To communicate with the uart port of the SoC you need a usb to serial converter:
+If you want to be able to program your device from an intel hex file, you will also need to install the following:
+
+* IntelHex: https://pypi.python.org/pypi/IntelHex
+* python-magic: https://pypi.python.org/pypi/python-magic
```
and 'pip install python-magic'

JelmerT commented on cc2538-bsl.py:
```diff
@@ -97,6 +104,62 @@ def mdebug(level, message, attr='\n'):
+
+        else:
+            mdebug(5, "Cannot auto-detect firmware filetype: Assuming .bin")
+            mdebug(10, "You will not be able to programme your device with a " \
```
Typo -> program? Not sure about this one. If it's correct British English, disregard this comment. ;)

JelmerT commented:
Sorry for the delay on these.

This PR seems to work great! Tested with 2.7 and 3.4. I made some small comments inline.

One thing I noticed: when you don't have the extra libs installed and you're trying to upload a hex file, it'll treat the file as a normal bin. If the file has the correct length it gets uploaded (but doesn't work of course), if not, you get an error about the size not being a multiple of 4. This behavior is to be expected, but there's never a hint to what is going wrong and how you can fix it (without activating verbose mode or looking at the readme).

Would it make sense to check the filename extension for `.hex`? And print the extra debug string in that case. Do most intel hex files use that extension? 
Bringing down the verbosity of the extra debug string from `10` to `5` would be another solution, but this clutters the normal output of the script of course.
Making a note in the `-h` might be another option.

Any ideas? This is not a big issue btw, I'm willing to pull as-is.

g-oikonomou commented:
@JelmerT thanks for the feedback. I've rebased this on top of current master and implemented a new logic:

* First, we try to auto-detect file type through python-magic. If this fails, fall-back to checking the extension and warn if running with `-V`.
* If the firmware is (or looks like) an Intel Hex, fail with a verbose error if IntelHex is not installed. Thus, we will no longer treat hex files as if they were a .bin.

I think the verb is always "program" in En-GB!

g-oikonomou commented:
Hmm hang on, this is buggy. Do _not_ merge

g-oikonomou commented:
There we go, ready for consideration now!

JelmerT commented:
Works like a charm!

This is very good stuff, thanks @g-oikonomou !
